deps: path-to-regexp@6.2.0 (#102)
pillarjs/router: HISTORY.md

> `2.0.0-beta.1` / 2020-03-29

Reviewer: These headers shouldn't have the backticks; it deviates from all the other history file formats and breaks our tooling.

Author: Sorry - couldn't resist the temptation. Reverted.

This incorporates all changes after 1.3.5 up to 1.3.6.

* deps: `path-to-regexp@6.2.0` (See [release changelog](https://github.com/pillarjs/path-to-regexp/releases/tag/v6.2.0))
  - Support for **named capturing groups** in paths using `RegExp`, i.e.: `/\/(?<group>.+)/` would result in `req.params.group` being populated.
  - Custom **prefix and suffix groups** using `{}`, e.g.: `/:entity{-:action}?` would match `/user` and `/user-delete`.

>   - **BREAKING**: Parenthesis and bracket literals now require escaping, i.e.: `/user(:id)` or `/user{:id}` no longer match `/user(42)`, `/user{42}` as before unless written as `/user\\(:id\\)` or `/user\\{:id\\}`.

Reviewer: I didn't test the `{}` one, since that was added in path-to-regexp 6, but I tested the `/user(:id)` one and `/user(:id)` doesn't match `/user(42)` in 2.0.0-beta.1 either, and needs the escapes. Was this change just for `{}` and was the `()` a typo? I cannot find this noted in `path-to-regexp` changelog, either.

Reviewer: I went ahead and tested `/user{:id}` in 2.0.0-beta.1 as well and it doesn't work with or without `\\` chars. Here is the test I was using to try it out:
```js
it('should match single path segment', function (done) {
  var router = new Router()
  var route = router.route('/user(:id)')
  var server = createServer(router)

  route.all(sendParams)

  request(server)
    .get('/user(1)')
    .expect(200, '{"id":"1"}', done)
})
```
It returns 404 for both 2.0.0-beta.1 release and on this PR. I'll add these as tests to the test suite as well, as they are good tests to have 👍

Author: I tested this using `path-to-regexp@6.2.0` itself, if I remember correctly. Yeah, without escaping brackets you'd get a `404`.

```js
var match = require("path-to-regexp").match

match('/user/\\(:id\\)')('/user/(42)')
// {path: "/user/(42)", index: 0, params: {id: "42"}}

m('/user/\\{:id\\}')('/user/{42}')
// {path: "/user/{42}", index: 0, params: {id: "42"}}
```

Reviewer: Right, I understand that. But I guess I wasn't clear in my point: I am saying that I don't think that _changed_ between 2.0.0-beta.1 and this PR, so it doesn't seem like it should be in our History as a breaking change, since nothing broke. For example, using the path-to-regexp that is in 2.0.0-beta.1 directly, as you tested directly with the one in this PR, here are the results on your tests:
```
$ node
Welcome to Node.js v16.7.0.
Type ".help" for more information.
> require("path-to-regexp/package").version
'3.2.0'
> match=require("path-to-regexp").match
[Function: match]
> match('/user/(:id)')('/user/(42)')
false
> match('/user/\\(:id\\)')('/user/(42)')
{ path: '/user/(42)', index: 0, params: { id: '42' } }
> match('/user/{:id}')('/user/{42}')
{ path: '/user/{42}', index: 0, params: { id: '42' } }
> match('/user/\\{:id\\}')('/user/{42}')
{ path: '/user/{42}', index: 0, params: { id: '42' } }
```

You can see that yes, it changes the `{:id}` one, but as I was saying, it doesn't seem to change the parenthesis one at all, right? So when it says "`/user(:id)` [...] no longer match `/user(42)`" in this update, how is that breaking, as it didn't work in the previous version, either.

Author: Ah, gotcha 👍🏼. Sorry, misunderstood what you were trying to say. Yeah, you're right. Maybe we could move that item about escaping brackets to the previous release changelog? I don't see it mentioned anywhere else and I think it should be.

Reviewer: Yes, if indeed it does work in router 1.x and then didn't work in router 2.0.0-beta.1 then it was a miss for that changelog and we should add it there. I haven't confirmed it as I am on mobile now, but if confirmed, one of us can update it.

Author: Looks like it did work in `path-to-regexp@0.1.7`.

```js
var pathToRegexp = require('path-to-regexp@0.1.7')
'/user/(42)'.match(pathToRegexp('/user/(:id)'))

// ['/user/(42)', '(42)']
```

Reviewer: Ok, cool. So then the entry does not belong in this change set, as it's just a missing one from 2.0.0-beta.1. It should at least be removed from this changeset. Ideally we should add it over retroactively to the 2.0.0-beta.1 changeset, but you don't need to do that in this PR if you don't want to.

2.0.0-beta.1 / 2020-03-29
=========================

This incorporates all changes after `1.3.3` up to `1.3.5`.

  * Internalize private `router.process_params` method
  * Remove `debug` dependency
  * deps: `array-flatten@3.0.0`
  * deps: `parseurl@~1.3.3`
  * **BREAKING**:  deps: `path-to-regexp@3.2.0`
    - Add new `?`, `*`, and `+` parameter modifiers.
    - Matching group expressions are only RegExp syntax.
      `(*)` is no longer valid and must be written as `(.*)`, for example.
    - Named matching groups no longer available by position in `req.params`.
      `/:foo(.*)` only captures as `req.params.foo` and not available as
      `req.params[0]`.
    - Regular expressions can only be used in a matching group.
      `/\\d+` is no longer valid and must be written as `/(\\d+)`.
    - Special `*` path segment behavior removed.
      `/foo/*/bar` will match a literal `*` as the middle segment.

>     - Parenthesis and bracket literals now require escaping, i.e.: `/user(:id)` or `/user{:id}` no longer match `/  user(42)`, `/user{42}` as before unless written as `/user\\(:id\\)` or `/user\\{:id\\}`.

Reviewer: I think this got messed up in the move. From my recollection from the conversation only the parenthesis stuff changed in this release; the curly brace stuff did change in this PR. Only the parans one should have been moved down here, right?

Author: @dougwilson I think you're right. Corrected it. Also rebased with latest `2.0`.

This incorporates all changes after 1.3.5 up to 1.3.6.

  * Add support for returned, rejected Promises to `router.param`
  * **BREAKING**: deps: `path-to-regexp@6.2.0` (See [release changelog](https://github.com/pillarjs/path-to-regexp/releases/tag/v6.2.0))
    - Support for **named capturing groups** in paths using `RegExp`, i.e.: `/\/(?<group>.+)/` would result in `req.params.group` being populated.
    - Custom **prefix and suffix groups** using `{}`, e.g.: `/:entity{-:action}?` would match `/user` and `/user-delete`.

>     - Unbalanced patterns now produce an error: `/test(foo` previously worked, but now it expects `(` to be escaped for the previous behavior, e.g. `/test\\(foo`.

Reviewer 2: The release notes referenced above says the previous behavior can be achieved with `/test\(foo` but here it states `/test\\(foo`. Is there a reason for the difference? Apologies if I missed something obvious!

Author: Hey @xjamundx 👋🏼 Thanks for chiming in.

You're absolute right, good catch. Yeah, I think `path-to-regexp` release notes are wrong, or misleading at the very least. `/test(foo` no longer works because it starts a capturing group that it's never closed - but if you want the literal `(` to match you'd need to escape it with '\'. This being JS, escaping non-special chars with `\` just echoes the char bar (i.e.: `'\('` -> `'('`). So in order to forward the escaped sequence to the underlying `RegExp`, you'd need to escape the backslash as well.

See it in action:

```js
> const { pathToRegexp } = require('path-to-regexp')

> pathToRegexp('/test\(foo').test('/test(foo')
Uncaught TypeError: Unbalanced pattern at 5

> pathToRegexp('/test\\(foo').test('/test(foo')
true
```

In short, I think `router`'s description makes more sense and should cause less confusion.

>     - Unbalanced patterns would now produce an error: `/test(foo` previously would worked, now it expects either `(` to be closed or the character to be escaped for the previous behavior, e.g. `/test\\(foo`.

Reviewer: ```suggestion
    - Unbalanced patterns now produces an error: `/test(foo` previously worked, but now it expects `(` to escaped for the previous behavior, e.g. `/test\\(foo`.
```

I think this wording may be better/clearer. It seems that only escaping it actually reproduces the previous behavior; closing the parans results in different behavior as it now created a capture group.

Author: I borrowed the wording straight out of [`path-to-regexp` release notes](https://github.com/pillarjs/path-to-regexp/releases/tag/v6.0.0).

Reviewer: Ok, but the English does not make sense as it is written here. If you have a better suggestion, please feel free to update it.

Author: @dougwilson Done. I went with your suggestion. Anything else?

Author: 👀

Reviewer 2: 🙏🏻

  * Fix OPTIONS responses to include the HEAD method properly
  * Remove support for leading colon in `router.param(name, fn)`
  * Use `array-flatten` for flattening arrays
  * deps: `debug@~2.1.1`

>   * deps: `methods@~1.1.1`

Reviewer: Can all these extra unrelated changes be cleaned up or do you want me to review them?

Author: Don't think there's much "to review". I don't care about them. I'll revert them.

Reviewer: Ok. Otherwise afaik this pr is good to go. I will merge it asap.

Author: Done. `HISTORY.md` should be clean.
